### PR TITLE
ref(sourcemaps): Redesign lookup of source and sourcemaps

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at implementing a new system for looking up source and sourcemaps files for each frame. This mechanism has been designed to account for the future implementation of debug ids.

#sync-getsentry

### CODE REVIEW

**Review by mitsuhiko — APPROVED** [review]

**Review by mitsuhiko — APPROVED** [review]
Minor nit: `encode()` should be `encode('utf-8')` as discussed.

File: src/sentry/lang/javascript/processor.py [hunk 1]
```diff
@@ -880,22 +875,34 @@ def __init__(self, *args, **kwargs):
...
-        # cache holding mangled code, original code, and errors associated with
-        # each abs_path in the stacktrace
-        self.cache = SourceCache()
+        # All the following dictionaries have been defined top level for simplicity reasons. Because this code will
+        # be ported to Symbolicator we wanted to keep it as simple and explicit as possible. This comment also
+        # applies to the many repetitions across the code.
+
+        # Cache holding the results of the fetching by url.
+        self.fetch_by_url_sourceviews = {}
+        self.fetch_by_url_errors = {}
```
Comment (Swatinem) [inline src/sentry/lang/javascript/processor.py:888]: is this the only place that stores (cached) download errors?

**Review by Swatinem — APPROVED** [review]
code and comments are a lot more readable!

This also gave me a couple of ideas for symbolicator. CC @kamilogorek @loewenheim just fyi

File: src/sentry/lang/javascript/processor.py (hunk 1 above)
Comment (iambriccardo) [reply src/sentry/lang/javascript/processor.py:888]: Yep, it is.

**Review by mitsuhiko — APPROVED** [review]

File: src/sentry/lang/javascript/processor.py [hunk 2]
```diff
@@ -853,6 +695,159 @@ def format_groups(p):
...
+    """
+
+    def __init__(self, project=None, release=None, dist=None, allow_scraping=True):
+        self.project = project
+        self.release = release
+        self.dist = dist
+        self.allow_scraping = allow_scraping
+
+    def bind_release(self, release=None, dist=None):
+        """Updates the fetcher with a project, release and dist."""
```
Comment (kamilogorek) [inline src/sentry/lang/javascript/processor.py:711]: It doesn't bind project the 👀

Comment (iambriccardo) [reply src/sentry/lang/javascript/processor.py:711]: @mitsuhiko what is your excuse? 👀🤣

**Pull request merged**